scripts/alert_emails_report/generate-report.ts

```diff
@@ -0,0 +1,134 @@
+  } catch (err) {
+    console.error('Error updating engagement stats', err);
+  }
+
+  console.info('Done.');
+  console.info(`Check the report: ${REPORT_URL}`);
+  process.exit(0);
+}
+
+async function updateSubscriptionsByLocation() {
+  const db = getFirestore();
+  const gsheets = new GoogleSheets(keyFile);
+
+  const subscriptions = await fetchAllAlertSubscriptions(db);
+  const groupedByFips = _.groupBy(subscriptions, fipsEmail => fipsEmail.fips);
+  const countByFips = _.map(groupedByFips, (items, fips) => {
+    return { fips, count: items.length };
+  });
+
+  const [stateCounts, countyCounts] = _.partition(countByFips, ({ fips }) =>
+    isStateFips(fips),
+  );
+  const countyStats = formatCountyStats(countyCounts);
+  const stateStats = formatStateStats(stateCounts);
+
+  const countyRange = 'subscriptions by county!A2:D';
+  await gsheets.clearAndAppend(SPREADSHEET_ID, countyRange, countyStats);
+
+  const stateRange = 'subscriptions by state!A2:B';
+  await gsheets.clearAndAppend(SPREADSHEET_ID, stateRange, stateStats);
+}
+
+function formatStateStats(stats: FipsCount[]): Cell[][] {
+  const data = stats.map(({ fips, count }) => [
+    getLocationNameForFips(fips),
+    count,
+  ]);
+  return _.sortBy(data, item => item[0]);
+}
+
+function formatCountyStats(stats: FipsCount[]): Cell[][] {
+  const data = stats.map(({ fips, count }) => {
+    const county = findCountyByFips(fips);
+    const fipsCode = `'${fips}`;
+    const countyName: string = county?.county
+      ? county?.county
+      : 'Unknown county';
+    const stateCode: string = county?.state_code ? county?.state_code : '-';
+    return [fipsCode, countyName, stateCode, count];
+  });
+  return _.sortBy(data, item => item[0]);
+}
+
+async function updateEngagementStats() {
+  const dateTo = new Date();
+  const dateFrom = moment().subtract(7, 'day').toDate();
+  const gsheets = new GoogleSheets(keyFile);
+  const range = 'data_engagement!A2:F';
+  const rows = await fetchEngagementStats(dateFrom, dateTo);
+  await gsheets.clearAndAppend(SPREADSHEET_ID, range, rows);
+}
+
+async function fetchEngagementStats(from: Date, to: Date): Promise<Cell[][]> {
+  const cm = new CampaignMonitor(process.env.CREATE_SEND_TOKEN);
+  const dateFrom = new Date('2020-07-23');
+  const dateTo = new Date();
+  const groups = await cm.fetchTransactionalGroups();
+
+  const alertGroupNames = groups
+    .filter(group => group.Group.startsWith(ALERT_EMAIL_GROUP_PREFIX))
+    .map(group => group.Group)
+    .sort();
+
+  const stats = await Promise.all(
+    alertGroupNames.map(async (groupName: string) => {
+      const groupStats = await cm.fetchTransactionalStats(
+        groupName,
+        dateFrom,
+        dateTo,
+      );
+      return formatGroupStats(groupName, groupStats);
+    }),
+  );
+
+  return stats;
+}
+
+function formatGroupStats(groupName: string, stats: CampaignMonitorStats) {
+  return [groupName, stats.Sent, stats.Opened, stats.Clicked, stats.Bounces];
```

Comment:
FWIW- I think converting opened and clicked to percentages might make sense, since they'll vary based on how many emails were sent.  So to compare, we need to "normalize" them.

I _think_ (but you might want to check) that in the CM UI they show:

Open rate = opened / sent
Click rate = clicked / opened

I would probably do the same here.

Alternatively (more flexible), maybe we can just set up the spreadsheet so that it calculates these automatically in the spreadsheet (as additional computed columns probably)?

Reply:
I think I'm going to set that up in Google Sheets - the formating is lost each time I write new data, so the plan is to have raw data sheets and then to use formulas to make the report nicer.

I checked, that's how CM calculates the rates.